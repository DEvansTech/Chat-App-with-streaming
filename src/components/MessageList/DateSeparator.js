import React from 'react';
import styled from '@stream-io/styled-components';
import PropTypes from 'prop-types';

import { themed } from '../../styles/theme';
import { withTranslationContext } from '../../context';

const Container = styled.View`
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  margin-top: 20;
  margin-bottom: 20;
  ${({ theme }) => theme.messageList.dateSeparator.container.css}
`;

const Line = styled.View`
  flex: 1;
  height: 0.5;
  background-color: ${({ theme }) => theme.colors.light};
  ${({ theme }) => theme.messageList.dateSeparator.line.css}
`;

const DateText = styled.Text`
  margin-left: 5;
  margin-right: 5;
  text-align: center;
  text-transform: uppercase;
  font-size: 10;
  opacity: 0.8;
  ${({ theme }) => theme.messageList.dateSeparator.dateText.css}
`;

const Date = styled.Text`
  font-weight: 700;
  font-size: 10;
  text-transform: uppercase;
  opacity: 0.8;
  ${({ theme }) => theme.messageList.dateSeparator.date.css}
`;

/**
 * @extends PureComponent
 * @example ../docs/DateSeparator.md
 */

class DateSeparator extends React.PureComponent {
  static propTypes = {
    message: PropTypes.object.isRequired,
    /**
     * Formatter function for date object.
     *
     * @param date Date object of message
     * @returns string
     */
    formatDate: PropTypes.func,
  };

  static themePath = 'messageList.dateSeparator';

  render() {
    const { message, formatDate, tDateTimeParser } = this.props;

    return (
      <Container>
        <Line />
        <DateText>
          {formatDate ? (
            formatDate(message.date)
          ) : (
            <React.Fragment>
              <Date>{tDateTimeParser(message.date).calendar()}</Date>
            </React.Fragment>
          )}
        </DateText>
        <Line />
      </Container>
    );
  }
}

export default withTranslationContext(themed(DateSeparator));
