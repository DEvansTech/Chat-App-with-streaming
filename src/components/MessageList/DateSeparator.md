The date separator between messages.
Here's what it looks like for today.

```js
const date = new Date();
const data = require('../docs/data');

<Chat client={data.client}>
  <DateSeparator {...data.translationContext} message={{ date }} />
  <DateSeparator
    {...data.translationContext}
    message={{ date }}
    alignment="center"
  />
  <DateSeparator
    {...data.translationContext}
    message={{ date }}
    alignment="left"
  />
</Chat>;
```

and for a date in the past:

```js
const data = require('../docs/data');

const date = new Date('December 17, 1995 03:24:00');
<Chat client={data.client}>
  <DateSeparator {...data.translationContext} message={{ date }} />
  <DateSeparator
    {...data.translationContext}
    message={{ date }}
    alignment="center"
  />
  <DateSeparator
    {...data.translationContext}
    message={{ date }}
    alignment="left"
  />
</Chat>;
```

and adding custom date formatting:

```js
const data = require('../docs/data');

const date = new Date('December 17, 1995 03:24:00');

function formatDate(d) {
  return <h2>Messages posted after {d.toDateString()}</h2>;
}

<Chat client={data.client}>
  <DateSeparator
    {...data.translationContext}
    formatDate={formatDate}
    message={{ date }}
  />
  <DateSeparator
    {...data.translationContext}
    formatDate={formatDate}
    message={{ date }}
    alignment="center"
  />
  <DateSeparator
    {...data.translationContext}
    formatDate={formatDate}
    message={{ date }}
    alignment="left"
  />
</Chat>;
```
