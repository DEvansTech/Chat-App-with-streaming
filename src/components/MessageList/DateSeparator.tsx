import React from 'react';

import type { DateSeparator as DateSeparatorType } from './utils/insertDates';

import {
  isDayOrMoment,
  TDateTimeParserInput,
  useTranslationContext,
} from '../../contexts/translationContext/TranslationContext';
import { styled } from '../../styles/styledComponents';
import { themed } from '../../styles/theme';

import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
} from '../../types/types';

const Container = styled.View`
  align-items: center;
  flex-direction: row;
  justify-content: center;
  margin-vertical: 20px;
  ${({ theme }) => theme.messageList.dateSeparator.container.css}
`;

const Date = styled.Text`
  font-size: 10px;
  font-weight: 700;
  opacity: 0.8;
  text-transform: uppercase;
  ${({ theme }) => theme.messageList.dateSeparator.date.css}
`;

const DateText = styled.Text`
  font-size: 10px;
  margin-horizontal: 5px;
  opacity: 0.8;
  text-align: center;
  text-transform: uppercase;
  ${({ theme }) => theme.messageList.dateSeparator.dateText.css}
`;

const Line = styled.View`
  background-color: ${({ theme }) => theme.colors.light};
  flex: 1;
  height: 0.5px;
  ${({ theme }) => theme.messageList.dateSeparator.line.css}
`;

export type DateSeparatorProps<
  At extends Record<string, unknown> = DefaultAttachmentType,
  Ch extends Record<string, unknown> = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends Record<string, unknown> = DefaultEventType,
  Me extends Record<string, unknown> = DefaultMessageType,
  Re extends Record<string, unknown> = DefaultReactionType,
  Us extends Record<string, unknown> = DefaultUserType
> = {
  message: DateSeparatorType<At, Ch, Co, Ev, Me, Re, Us>;
  /**
   * Formatter function for date object.
   *
   * @param date TDateTimeParserInput object of message
   * @returns string
   */
  formatDate?: (date: TDateTimeParserInput) => string;
};

/**
 * @example ./DateSeparator.md
 */
const DateSeparator = <
  At extends Record<string, unknown> = DefaultAttachmentType,
  Ch extends Record<string, unknown> = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends Record<string, unknown> = DefaultEventType,
  Me extends Record<string, unknown> = DefaultMessageType,
  Re extends Record<string, unknown> = DefaultReactionType,
  Us extends Record<string, unknown> = DefaultUserType
>({
  formatDate,
  message,
}: DateSeparatorProps<At, Ch, Co, Ev, Me, Re, Us>) => {
  const { tDateTimeParser } = useTranslationContext();

  let date;

  if (formatDate) {
    date = formatDate(message.date as TDateTimeParserInput);
  } else {
    date = tDateTimeParser(message.date as TDateTimeParserInput);
  }

  return (
    <Container testID='date-separator'>
      <Line />
      <DateText>
        {formatDate ? (
          date
        ) : (
          <Date>{isDayOrMoment(date) ? date.calendar() : date}</Date>
        )}
      </DateText>
      <Line />
    </Container>
  );
};

DateSeparator.themePath = 'messageList.dateSeparator';

export default themed(DateSeparator) as typeof DateSeparator;
