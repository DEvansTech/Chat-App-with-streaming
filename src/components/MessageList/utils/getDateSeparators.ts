import type { PaginatedMessageListContextValue } from '../../../contexts/paginatedMessageListContext/PaginatedMessageListContext';
import type { ThreadContextValue } from '../../../contexts/threadContext/ThreadContext';
import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../../types/types';

export type GetDateSeparatorsParams<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
> = {
  messages:
    | PaginatedMessageListContextValue<At, Ch, Co, Ev, Me, Re, Us>['messages']
    | ThreadContextValue<At, Ch, Co, Ev, Me, Re, Us>['threadMessages'];
  hideDateSeparators?: boolean;
};

export type DateSeparators = {
  [key: string]: Date;
};

export const getDateSeparators = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>(
  params: GetDateSeparatorsParams<At, Ch, Co, Ev, Me, Re, Us>,
) => {
  const { hideDateSeparators, messages } = params;
  const dateSeparators: DateSeparators = {};

  if (hideDateSeparators) {
    return dateSeparators;
  }

  for (let i = 0; i < messages.length; i++) {
    const previousMessage = messages[i - 1];
    const message = messages[i];

    if (
      message.type === 'channel.event' ||
      message.type === 'message.read' ||
      message.deleted_at
    ) {
      continue;
    }

    const messageDate = message.created_at.getDay();

    const prevMessageDate = previousMessage
      ? previousMessage.created_at.getDay()
      : messageDate;

    if (i === 0 || messageDate !== prevMessageDate) {
      dateSeparators[message.id] = message.created_at;
    }
  }

  return dateSeparators;
};
