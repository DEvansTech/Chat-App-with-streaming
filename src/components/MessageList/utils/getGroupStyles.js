export const getGroupStyles = (messagesWithDates, noGroupByUser) => {
  const numberOfMessages = messagesWithDates.length;
  const messageGroupStyles = {};

  const messages = [...messagesWithDates];

  for (let i = 0; i < numberOfMessages; i++) {
    const previousMessage = messages[i - 1];
    const message = messages[i];
    const nextMessage = messages[i + 1];
    const groupStyles = [];

    if (message.type === 'channel.event') {
      continue;
    }

    if (message.type === 'message.date') {
      continue;
    }

    const userId = message.user ? message.user.id : null;

    const isTopMessage =
      !previousMessage ||
      previousMessage.type === 'message.date' ||
      previousMessage.type === 'system' ||
      previousMessage.type === 'channel.event' ||
      (previousMessage.attachments &&
        previousMessage.attachments.length !== 0) ||
      userId !== previousMessage.user.id ||
      previousMessage.type === 'error' ||
      previousMessage.deleted_at;

    const isBottomMessage =
      !nextMessage ||
      nextMessage.type === 'message.date' ||
      nextMessage.type === 'system' ||
      nextMessage.type === 'channel.event' ||
      (nextMessage.attachments && nextMessage.attachments.length !== 0) ||
      userId !== nextMessage.user.id ||
      nextMessage.type === 'error' ||
      nextMessage.deleted_at;

    if (isTopMessage) {
      groupStyles.push('top');
    }

    if (isBottomMessage) {
      if (isTopMessage || message.deleted_at || message.type === 'error') {
        groupStyles.splice(0, groupStyles.length);
        groupStyles.push('single');
      } else {
        groupStyles.push('bottom');
      }
    }

    if (!isTopMessage && !isBottomMessage) {
      if (message.deleted_at || message.type === 'error') {
        groupStyles.splice(0, groupStyles.length);
        groupStyles.push('single');
      } else {
        groupStyles.splice(0, groupStyles.length);
        groupStyles.push('middle');
      }
    }

    if (message.attachments.length !== 0) {
      groupStyles.splice(0, groupStyles.length);
      groupStyles.push('single');
    }

    if (noGroupByUser) {
      groupStyles.splice(0, groupStyles.length);
      groupStyles.push('single');
    }

    messageGroupStyles[message.id] = groupStyles;
  }

  return messageGroupStyles;
};
