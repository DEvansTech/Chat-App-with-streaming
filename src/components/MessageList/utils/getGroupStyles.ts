import type { DateSeparators } from './getDateSeparators';

import type { GroupType } from '../hooks/useMessageList';

import type { PaginatedMessageListContextValue } from '../../../contexts/paginatedMessageListContext/PaginatedMessageListContext';
import type { ThreadContextValue } from '../../../contexts/threadContext/ThreadContext';
import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../../types/types';
import type { FormatMessageResponse } from 'stream-chat';

export type GetGroupStylesParams<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
> = {
  dateSeparators: DateSeparators;
  messages:
    | PaginatedMessageListContextValue<At, Ch, Co, Ev, Me, Re, Us>['messages']
    | ThreadContextValue<At, Ch, Co, Ev, Me, Re, Us>['threadMessages'];
  hideDateSeparators?: boolean;
  maxTimeBetweenGroupedMessages?: number;
  noGroupByUser?: boolean;
  userId?: string;
};

export const getGroupStyles = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>(
  params: GetGroupStylesParams<At, Ch, Co, Ev, Me, Re, Us>,
) => {
  const {
    dateSeparators,
    hideDateSeparators,
    maxTimeBetweenGroupedMessages,
    messages,
    noGroupByUser,
    userId,
  } = params;
  const messageGroupStyles: { [key: string]: GroupType[] } = {};

  const messagesFilteredForNonUser = messages.filter(
    (message) => !message.deleted_at || userId === message.user?.id,
  );

  for (let i = 0; i < messagesFilteredForNonUser.length; i++) {
    const previousMessage = messagesFilteredForNonUser[i - 1] as
      | typeof messagesFilteredForNonUser[0]
      | undefined;
    const message = messagesFilteredForNonUser[i];
    const nextMessage = messagesFilteredForNonUser[i + 1] as
      | typeof messagesFilteredForNonUser[0]
      | undefined;
    const groupStyles: GroupType[] = [];

    const userId = message?.user?.id || null;

    const isTopMessage =
      !previousMessage ||
      previousMessage.type === 'system' ||
      userId !== previousMessage?.user?.id ||
      previousMessage.type === 'error' ||
      !!previousMessage.deleted_at ||
      (!hideDateSeparators && dateSeparators[message.id]) ||
      messageGroupStyles[previousMessage.id]?.includes('bottom');

    const isBottomMessage =
      !nextMessage ||
      nextMessage.type === 'system' ||
      userId !== nextMessage?.user?.id ||
      nextMessage.type === 'error' ||
      !!nextMessage.deleted_at ||
      (!hideDateSeparators && dateSeparators[nextMessage.id]) ||
      (maxTimeBetweenGroupedMessages !== undefined &&
        nextMessage.created_at.getTime() - message.created_at.getTime() >
          maxTimeBetweenGroupedMessages);

    /**
     * Add group styles key for top message
     */
    if (isTopMessage) {
      groupStyles.push('top');
    }

    /**
     * Add group styles key for bottom message
     */
    if (isBottomMessage) {
      /**
       * If the bottom message is also the top, or deleted, or an error,
       * add the key for single message instead of bottom
       */
      if (isTopMessage || message.deleted_at || message.type === 'error') {
        groupStyles.splice(0, groupStyles.length);
        groupStyles.push('single');
      } else {
        groupStyles.push('bottom');
      }
    }

    /**
     * Add the key for all non top or bottom messages, if the message is
     * deleted or an error add the key for single otherwise middle
     */
    if (!isTopMessage && !isBottomMessage) {
      if (message.deleted_at || message.type === 'error') {
        groupStyles.splice(0, groupStyles.length);
        groupStyles.push('single');
      } else {
        groupStyles.splice(0, groupStyles.length);
        groupStyles.push('middle');
      }
    }

    /**
     * If noGroupByUser set add the key for single
     */
    if (noGroupByUser) {
      groupStyles.splice(0, groupStyles.length);
      groupStyles.push('single');
    }

    if (message.id) {
      messageGroupStyles[message.id] = groupStyles;
    }
  }

  return messageGroupStyles;
};
