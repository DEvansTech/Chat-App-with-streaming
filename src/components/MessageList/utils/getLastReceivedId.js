export const getLastReceivedId = (messages) => {
  let lastReceivedId = null;

  for (let i = messages.length; i > 0; i--) {
    if (
      messages[i] !== undefined &&
      messages[i].status !== undefined &&
      messages[i].status === 'received'
    ) {
      lastReceivedId = messages[i].id;
      break;
    }
  }

  return lastReceivedId;
};
