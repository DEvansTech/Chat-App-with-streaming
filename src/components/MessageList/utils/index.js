export { getGroupStyles } from './getGroupStyles';
export { getLastReceivedMessage } from './getLastReceivedMessage';
export { getReadStates } from './getReadStates';
export { insertDates } from './insertDates';
