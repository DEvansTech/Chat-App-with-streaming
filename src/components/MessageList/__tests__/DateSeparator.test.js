import React from 'react';
import { cleanup, render, waitFor } from '@testing-library/react-native';

import {
  generateMessage,
  generateStaticMessage,
} from 'mock-builders/generator/message';
import { generateStaticUser, generateUser } from 'mock-builders/generator/user';

import DateSeparator from '../DateSeparator';

import { TranslationProvider } from '../../../contexts/translationContext/TranslationContext';
import { Streami18n } from '../../../utils/Streami18n';

afterEach(cleanup);

describe('DateSeparator', () => {
  it('should render date separator', async () => {
    const i18nInstance = new Streami18n();
    const translators = await i18nInstance.getTranslators();
    const user = generateUser();
    const message = generateMessage({ user });
    const { queryByTestId } = render(
      <TranslationProvider value={translators}>
        <DateSeparator message={message} />
      </TranslationProvider>,
    );

    await waitFor(() => {
      expect(queryByTestId('date-separator')).toBeTruthy();
    });
  });

  it('should render date separator with custom date format', async () => {
    const i18nInstance = new Streami18n();
    const translators = await i18nInstance.getTranslators();
    const user = generateUser();
    const message = generateMessage({ user });
    const { getByText } = render(
      <TranslationProvider value={translators}>
        <DateSeparator
          formatDate={(date) => date}
          message={{ ...message, date: 'Hello World' }}
        />
      </TranslationProvider>,
    );

    await waitFor(() => {
      expect(getByText('Hello World')).toBeTruthy();
    });
  });

  it('should match date separator snapshot', async () => {
    const i18nInstance = new Streami18n();
    const translators = await i18nInstance.getTranslators();
    const user = generateStaticUser(0);
    const message = generateStaticMessage('Hello World', { user });
    const { toJSON } = render(
      <TranslationProvider value={translators}>
        <DateSeparator message={{ ...message, date: message.created_at }} />
      </TranslationProvider>,
    );

    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
    });
  });
});
