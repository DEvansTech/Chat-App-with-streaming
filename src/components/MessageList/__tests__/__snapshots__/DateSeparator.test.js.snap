// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateSeparator should match date separator snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "center",
        "marginBottom": 20,
        "marginTop": 20,
      },
    ]
  }
  testID="date-separator"
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#EBEBEB",
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "height": 0.5,
        },
      ]
    }
  />
  <Text
    style={
      Array [
        Object {
          "fontSize": 10,
          "marginLeft": 5,
          "marginRight": 5,
          "opacity": 0.8,
          "textAlign": "center",
          "textTransform": "uppercase",
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontSize": 10,
            "fontWeight": "700",
            "opacity": 0.8,
            "textTransform": "uppercase",
          },
        ]
      }
    >
      04/27/2020
    </Text>
  </Text>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#EBEBEB",
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "height": 0.5,
        },
      ]
    }
  />
</View>
`;
